src/common.c

```diff
@@ -84,73 +84,93 @@ void replyRMCallError(RedisModuleCtx *ctx, int err, const char *cmd, size_t len)
 }
 
 /* Create a -MOVED reply. */
-void replyRedirect(RedisModuleCtx *ctx, int slot, NodeAddr *addr)
+void replyRedirect(RedisModuleCtx *ctx, unsigned int slot, NodeAddr *addr)
 {
     char buf[sizeof(addr->host) + 256];
 
-    snprintf(buf, sizeof(buf), "MOVED %d %s:%u", slot, addr->host, addr->port);
+    snprintf(buf, sizeof(buf), "MOVED %u %s:%u", slot, addr->host, addr->port);
     RedisModule_ReplyWithError(ctx, buf);
 }
 
-static const char *err_clusterdown = "CLUSTERDOWN No raft leader";
+/* Create a -ASK reply. */
+void replyAsk(RedisRaftCtx *rr, RedisModuleCtx *ctx, unsigned int slot) {
+    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
+    if (!sg) {
+        RedisModule_ReplyWithError(ctx, "ERR no importing shard group to ask");
+        return;
+    }
+
+    char buf[sizeof(sg->nodes[0].addr.host) + 256];
+    snprintf(buf, sizeof(buf), "ASK %u %s:%u", slot, sg->nodes[0].addr.host, sg->nodes[0].addr.port);
+    RedisModule_ReplyWithError(ctx, buf);
+}
+
+/* Create a CROSSSLOT response */
+void replyCrossSlot(RedisModuleCtx *ctx)
+{
+    RedisModule_ReplyWithError(ctx, "CROSSSLOT Keys in request don't hash to the same slot");
+}
+
+void replyClusterDownWithNoRaftLeader(RedisModuleCtx *ctx)
+{
+    RedisModule_ReplyWithError(ctx, "CLUSTERDOWN No raft leader");
+}
+
+void replyWithFormatErrorString(RedisModuleCtx *ctx, const char * fmt, ...)
+{
+    va_list ap;
+
+    va_start(ap, fmt);
+    int len = vsnprintf(NULL, 0, fmt, ap);
+    va_end(ap);
+
+    char * buf = RedisModule_Alloc(len+1);
+
+    va_start(ap, fmt);
+    vsnprintf(buf, len, fmt, ap);
+    va_end(ap);
+
+    buf[len] = '\0';
+    RedisModule_ReplyWithError(ctx, buf);
+    RedisModule_Free(buf);
+
+}
```

Comment:
```suggestion
void replyError(RedisModuleCtx *ctx, const char *fmt, ...)
{
    char buf[512];
    va_list ap;

    va_start(ap, fmt);
    vsnprintf(buf, sizeof(buf), fmt, ap);
    va_end(ap);

    RedisModule_ReplyWithError(ctx, buf);
}
```
I just have a couple of nitpicks:
- Shorter function name is easier to read
- I don't know if we ever going to have an error longer than, say 512 bytes. Even if we have, probably it's okay to have it truncated. So, I'd go with static buffer here. 

If you don't like these changes, please fix pointer indentations and delete the empty line at the end. Also, just fyi, vsnprintf() will always write `\0` at the end, no need it add it separately. (but I know, it feels safer this way :) )